plutus-playground/plutus-playground-server/app/Auth.hs

```diff
@@ -0,0 +1,315 @@
+          , Just $ encodeUtf8 $ _configRedirectUrl <> _configRedirectEndpoint)
+        , ("scope", Just oauthScopes)
+        , ("client_id", Just $ encodeUtf8 _configGithubClientId)
+        ] $
+      _githubEndpointsAuthLocation
+    oauthScopes = "gist"
+
+twoWeeks :: NominalDiffTime
+twoWeeks = 60 * 60 * 24 * 7 * 2
+
+authStatus ::
+     (MonadNow m, MonadLogger m) => Config -> Maybe Text -> m AuthStatus
+authStatus Config {..} cookieHeader = do
+  now <- getPOSIXTime
+  _authStatusAuthRole <-
+    case extractGithubToken _configJWTSignature now cookieHeader of
+      Right _ -> pure GithubUser
+      Left err -> do
+        logErrorN $
+          "Failed to extract github token at step: " <> Text.pack (show err)
+        pure Anonymous
+  pure AuthStatus {..}
+
+extractGithubToken ::
+     JWT.Signer -> POSIXTime -> Maybe Text -> Either Text (Token 'Github)
+extractGithubToken signer now cookieHeader =
+  runTrace $ do
+    attempt "Reading cookies."
+    cookies <- parseCookies . encodeUtf8 <$> withTrace cookieHeader
+    attempt $ "Looking for Session ID cookie: " <> Text.pack (show cookies)
+    githubAuth <- withTrace $ lookup (encodeUtf8 hSessionIdCookie) cookies
+    attempt $ "Reading JWT Cookie: " <> decodeUtf8 githubAuth
+    unverifiedJwt <- withTrace . JWT.decode . decodeUtf8 $ githubAuth
+    attempt "Verifying JWT Cookie."
+    verifiedJwt <- withTrace $ JWT.verify signer unverifiedJwt
+    let claims = JWT.claims verifiedJwt
+    attempt "Checking expiry date is set."
+    expiry <- withTrace $ JWT.exp claims
+    attempt "Checking expiry date is valid."
+    guard (JWT.secondsSinceEpoch expiry > now)
+    attempt "Looking for Github token claim."
+    json <-
+      withTrace .
+      Map.lookup githubTokenClaim . JWT.unClaimsMap . JWT.unregisteredClaims $
+      claims
+    attempt $ "Extracting token as a string:" <> Text.pack (show json)
+    withTrace $
+      case json of
+        String token -> pure $ Token token
+        _            -> Nothing
+
+githubCallback ::
+     (MonadLogger m, MonadWeb m, MonadError ServantErr m, MonadNow m)
+  => GithubEndpoints
+  -> Config
+  -> Maybe OAuthCode
+  -> m (Headers '[ Header "Set-Cookie" SetCookie, Header "Location" Text] NoContent)
+githubCallback _ _ Nothing =
+  with500Err . pure . Left $
+  "Expected a response from Github with an authorization code. Didn't get one!"
+githubCallback githubEndpoints config@Config {..} (Just code) = do
+  logInfoN $
+    "OAuth Code is: " <> Text.pack (show code) <>
+    ". Swapping for a long-lived token."
+  manager <- makeManager
+  let tokenRequest = makeTokenRequest githubEndpoints config code
+  logInfoN $ "Request: " <> Text.pack (show tokenRequest)
+  logInfoN $
+    "Request: " <> Text.pack (show (getRequestQueryString tokenRequest))
+  response <- with500Err $ doRequest tokenRequest manager
+  unless
+    (statusIsSuccessful (responseStatus response))
+    (with500Err . pure . Left $ "Response: " <> Text.pack (show response))
```

Comment:
how about `showText = Text.pack . show` somewhere, these are annoying, especially with all the parens they create

Reply:
Yes. Fixed. 👍